lib/ide/ast/impl/src/lib.rs

```diff
@@ -588,158 +582,197 @@ pub enum MacroPatternMatchRaw<T> {
-    /// Length of the textual representation of This type in Unicode codepoints.
-    ///
-    /// Usually implemented together with `HasSpan`.For any `T:HasSpan+HasRepr`
-    /// for `t:T` the following must hold: `t.span() == t.repr().len()`.
-    fn span(&self) -> usize;
+// === Tokenizer ===
+
+/// An enum of valid Ast tokens.I
+#[derive(Debug)]
+pub enum Token<'a> { Off(usize), Chr(char), Str(&'a str), Ast(&'a Ast) }
+
+/// Things that can be turned into stream of tokens.
+pub trait HasTokens {
+    /// Feeds TokenBuilder with stream of tokens obtained from `self`.
+    fn feed(&self, consumer:&mut impl TokenConsumer);
 }
 
-/// Counts codepoints.
-impl HasSpan for char {
-    fn span(&self) -> usize {
-        1
-    }
+/// Helper trait for Tokenizer, which consumes the token stream.
+pub trait TokenConsumer {
+    /// consumes one token
+    fn feed(&mut self, val:Token);
 }
 
-/// Counts codepoints.
-impl HasSpan for String {
-    fn span(&self) -> usize {
-        self.as_str().span()
+
+impl HasTokens for &str {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Str(self));
     }
 }
 
-/// Counts codepoints.
-impl HasSpan for &str {
-    fn span(&self) -> usize {
-        self.chars().count()
+impl HasTokens for String {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Str(self.as_str()));
     }
 }
 
-impl<T: HasSpan> HasSpan for Option<T> {
-    fn span(&self) -> usize {
-        self.as_ref().map_or(0, |wrapped| wrapped.span())
+impl HasTokens for usize {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Off(*self));
     }
 }
 
-impl<T: HasSpan> HasSpan for Vec<T> {
-    fn span(&self) -> usize {
-        let spans = self.iter().map(|elem| elem.span());
-        spans.sum()
+impl HasTokens for char {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Chr(*self));
     }
 }
 
-impl<T: HasSpan> HasSpan for Rc<T> {
-    fn span(&self) -> usize {
-        self.deref().span()
+impl HasTokens for Ast {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Ast(self));
     }
 }
 
-impl<T: HasSpan, U: HasSpan> HasSpan for (T,U) {
-    fn span(&self) -> usize {
-        self.0.span() + self.1.span()
+impl<T:HasTokens> HasTokens for Option<T> {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        for t in self { t.feed(consumer); }
     }
 }
-impl<T: HasSpan, U: HasSpan, V: HasSpan> HasSpan for (T,U,V) {
-    fn span(&self) -> usize {
-        self.0.span() + self.1.span() + self.2.span()
+
+impl<T:HasTokens> HasTokens for Vec<T> {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        for t in self { t.feed(consumer); }
     }
 }
-impl HasSpan for usize {
-    fn span(&self) -> usize {
-        *self
+
+impl<T:HasTokens> HasTokens for Rc<T> {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        self.deref().feed(consumer);
```

Comment:
This line is very error prone. Do you know that after removing constraint `T:HasTokens` you will have hefre infinite loop? Use `.content()` instead of `.deref()` - and read docs agbout `content` to learn more why